src/data/chats/model_loader.rs

```diff
@@ -1,88 +1,199 @@
-                        load_model_tx.send(Err(err)).unwrap();
+#[derive(Default)]
+struct ModelLoaderInner {
+    status: ModelLoaderStatus,
+    file_id: Option<FileID>,
+}
+
+/// Unit for handling the non-blocking loading of models across threads.
+#[derive(Clone, Default)]
+pub struct ModelLoader(Arc<Mutex<ModelLoaderInner>>);
+
+impl ModelLoader {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn load_blocking(
+        &mut self,
+        file_id: FileID,
+        command_sender: Sender<Command>,
+    ) -> Result<(), ModelLoaderError> {
+        match self.status() {
+            ModelLoaderStatus::Loading => {
+                return Err(anyhow!("ModelLoader is already loading a model").into());
+            }
+            ModelLoaderStatus::Loaded => {
+                if let Some(prev_file_id) = self.file_id() {
+                    if prev_file_id == file_id {
+                        return Ok(());
                     }
                 }
-            } else {
-                load_model_tx.send(
-                    Err(anyhow::anyhow!("Error loading model"))
-                ).unwrap();
             }
+            _ => {}
+        };
 
-            SignalToUI::set_ui_signal();
-        });
-    }
+        self.set_status(ModelLoaderStatus::Loading);
+        self.set_file_id(Some(file_id.clone()));
+
+        let response = dispatch_load_command(command_sender, file_id.clone()).recv();
 
-    pub fn check_load_response(&mut self) -> Result<()> {
-        for msg in self.load_receiver.try_iter() {
-            match msg {
-                Ok(_) => {
-                    self.complete = true;
-                    return Ok(())
+        let result = if let Ok(response) = response {
+            match response {
+                Ok(LoadModelResponse::Completed(_)) => {
+                    self.set_status(ModelLoaderStatus::Loaded);
+                    Ok(())
+                }
+                Ok(response) => {
+                    let msg = format!("Unexpected response: {:?}", response);
+                    let err = ModelLoaderError::from(msg);
+                    self.set_status(ModelLoaderStatus::Failed(err.clone()));
+                    Err(err)
                 }
                 Err(err) => {
-                    self.complete = true;
-                    return Err(err.into())
+                    let err = ModelLoaderError::from(err);
+                    self.set_status(ModelLoaderStatus::Failed(err.clone()));
+                    Err(err)
                 }
             }
+        } else {
+            let err = ModelLoaderError::from("Internal communication error");
+            self.set_status(ModelLoaderStatus::Failed(err.clone()));
+            Err(err)
         };
 
-        Ok(())
+        SignalToUI::set_ui_signal();
+        result
     }
-}
\ No newline at end of file
+
+    pub fn load(&mut self, file_id: FileID, command_sender: Sender<Command>) {
```

Comment:
Provided just to have a function with the previous behavior although, forcing the caller yo be explicit about spawning a thread would be also an option.